HDFS-13694. Making md5 computing being in parallel with image loading.

leosunli opened the pull request:
Signed-off-by: sunlisheng <sunlisheng@xiaomi.com>

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 25 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1179 | trunk passed |
| +1 | compile | 63 | trunk passed |
| +1 | checkstyle | 45 | trunk passed |
| +1 | mvnsite | 72 | trunk passed |
| +1 | shadedclient | 811 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 50 | trunk passed |
| 0 | spotbugs | 176 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 174 | hadoop-hdfs-project/hadoop-hdfs in trunk has 1 extant findbugs warnings. |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 64 | the patch passed |
| +1 | compile | 60 | the patch passed |
| +1 | javac | 60 | the patch passed |
| -0 | checkstyle | 38 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 1 new + 28 unchanged - 0 fixed = 29 total (was 28) |
| +1 | mvnsite | 71 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 765 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 49 | the patch passed |
| -1 | findbugs | 186 | hadoop-hdfs-project/hadoop-hdfs generated 1 new + 1 unchanged - 0 fixed = 2 total (was 1) |
||| _ Other Tests _ |
| -1 | unit | 5199 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 60 | The patch does not generate ASF License warnings. |
| | | 8978 | |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-hdfs-project/hadoop-hdfs |
|  |  Should org.apache.hadoop.hdfs.server.namenode.FSImageFormatProtobuf$Loader$DigestThread be a _static_ inner class?  At FSImageFormatProtobuf.java:inner class?  At FSImageFormatProtobuf.java:[lines 176-203] |
| Failed junit tests | hadoop.hdfs.server.diskbalancer.TestDiskBalancer |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| JIRA Issue | HDFS-13694 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 7727bfa4491b 4.4.0-143-generic #169~14.04.2-Ubuntu SMP Wed Feb 13 15:00:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b76b843 |
| Default Java | 1.8.0_212 |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/1/artifact/out/branch-findbugs-hadoop-hdfs-project_hadoop-hdfs-warnings.html |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/1/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/1/artifact/out/new-findbugs-hadoop-hdfs-project_hadoop-hdfs.html |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/1/testReport/ |
| Max. process+thread count | 3517 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 29 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1036 | trunk passed |
| +1 | compile | 61 | trunk passed |
| +1 | checkstyle | 38 | trunk passed |
| +1 | mvnsite | 64 | trunk passed |
| +1 | shadedclient | 722 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 47 | trunk passed |
| 0 | spotbugs | 172 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 170 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 61 | the patch passed |
| +1 | compile | 57 | the patch passed |
| +1 | javac | 57 | the patch passed |
| -0 | checkstyle | 37 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 1 new + 28 unchanged - 0 fixed = 29 total (was 28) |
| +1 | mvnsite | 62 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 721 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 47 | the patch passed |
| +1 | findbugs | 175 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 4991 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 36 | The patch does not generate ASF License warnings. |
| | | 8408 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |
|   | hadoop.hdfs.server.balancer.TestBalancerWithNodeGroup |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.server.diskbalancer.TestDiskBalancer |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.qjournal.server.TestJournalNodeSync |
|   | hadoop.hdfs.server.namenode.snapshot.TestSnapshot |
|   | hadoop.tracing.TestTracing |
|   | hadoop.hdfs.server.namenode.snapshot.TestSnapRootDescendantDiff |
|   | hadoop.hdfs.server.balancer.TestBalancerRPCDelay |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| JIRA Issue | HDFS-13694 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 86b570d993b9 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 041e7a7 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/2/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/2/testReport/ |
| Max. process+thread count | 4599 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 28 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1053 | trunk passed |
| +1 | compile | 58 | trunk passed |
| +1 | checkstyle | 39 | trunk passed |
| +1 | mvnsite | 63 | trunk passed |
| +1 | shadedclient | 719 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 53 | trunk passed |
| 0 | spotbugs | 170 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 167 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 67 | the patch passed |
| +1 | compile | 58 | the patch passed |
| +1 | javac | 58 | the patch passed |
| -0 | checkstyle | 37 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 1 new + 28 unchanged - 0 fixed = 29 total (was 28) |
| +1 | mvnsite | 64 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 737 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 48 | the patch passed |
| -1 | findbugs | 174 | hadoop-hdfs-project/hadoop-hdfs generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
||| _ Other Tests _ |
| -1 | unit | 6193 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 40 | The patch does not generate ASF License warnings. |
| | | 9648 | |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-hdfs-project/hadoop-hdfs |
|  |  Should org.apache.hadoop.hdfs.server.namenode.FSImageFormatProtobuf$Loader$DigestThread be a _static_ inner class?  At FSImageFormatProtobuf.java:inner class?  At FSImageFormatProtobuf.java:[lines 176-203] |
| Failed junit tests | hadoop.hdfs.server.diskbalancer.TestDiskBalancer |
|   | hadoop.hdfs.TestWriteReadStripedFile |
|   | hadoop.hdfs.TestHdfsAdmin |
|   | hadoop.hdfs.TestErasureCodingPolicyWithSnapshotWithRandomECPolicy |
|   | hadoop.hdfs.TestMaintenanceState |
|   | hadoop.hdfs.tools.TestDFSZKFailoverController |
|   | hadoop.hdfs.TestDFSStripedOutputStreamWithRandomECPolicy |
|   | hadoop.hdfs.TestDistributedFileSystemWithECFileWithRandomECPolicy |
|   | hadoop.hdfs.TestErasureCodingPoliciesWithRandomECPolicy |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.TestRollingUpgradeDowngrade |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| JIRA Issue | HDFS-13694 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 2d2ecc60b973 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 041e7a7 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/3/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/3/artifact/out/new-findbugs-hadoop-hdfs-project_hadoop-hdfs.html |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/3/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/3/testReport/ |
| Max. process+thread count | 3937 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
     void load(File file) throws IOException {
       long start = Time.monotonicNow();
-      imgDigest = MD5FileUtils.computeMd5ForFile(file);
+      FSImageFormatProtobuf.Loader.DigestThread dt =
```
Can you just use DigestThread? Instead of the full declaration.

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
+      public void run() {
+        try {
+          digest = MD5FileUtils.computeMd5ForFile(file);
+        } catch (Throwable t) {
```
} catch (IOException ioe) {
  throw ioe;
} catch (Throwable t) {
  throw new IOException(t);
}

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
       return ctx;
     }
 
+    private class DigestThread extends Thread {
```
I think the common approach is to define this as static.

goiri replied:
Javadoc mentioning that this a thread for parallel MD5 computing to increase performance when loading.

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
+        try {
+          dt.join();
+          imgDigest = dt.getDigest();
+          if (dt.getException() != null) {
```
I guess you can move this to the previous line?

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
+          if (t instanceof IOException) {
+            ioe = (IOException) t;
+          } else {
+            ioe = new IOException(t);
```
Do we have unit tests throwing these exceptions?

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 30 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1077 | trunk passed |
| +1 | compile | 61 | trunk passed |
| +1 | checkstyle | 45 | trunk passed |
| +1 | mvnsite | 69 | trunk passed |
| +1 | shadedclient | 790 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 55 | trunk passed |
| 0 | spotbugs | 161 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 159 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 60 | the patch passed |
| +1 | compile | 56 | the patch passed |
| +1 | javac | 56 | the patch passed |
| -0 | checkstyle | 41 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 1 new + 28 unchanged - 0 fixed = 29 total (was 28) |
| +1 | mvnsite | 64 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 733 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 49 | the patch passed |
| -1 | findbugs | 169 | hadoop-hdfs-project/hadoop-hdfs generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
||| _ Other Tests _ |
| -1 | unit | 4826 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 8369 | |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-hdfs-project/hadoop-hdfs |
|  |  Should org.apache.hadoop.hdfs.server.namenode.FSImageFormatProtobuf$Loader$DigestThread be a _static_ inner class?  At FSImageFormatProtobuf.java:inner class?  At FSImageFormatProtobuf.java:[lines 176-203] |
| Failed junit tests | hadoop.hdfs.TestDatanodeDeath |
|   | hadoop.hdfs.server.datanode.TestDataNodeHotSwapVolumes |
|   | hadoop.hdfs.server.blockmanagement.TestBlockStatsMXBean |
|   | hadoop.hdfs.TestReconstructStripedFileWithRandomECPolicy |
|   | hadoop.hdfs.server.diskbalancer.TestDiskBalancer |
|   | hadoop.hdfs.TestDFSStripedOutputStream |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux a71b8f184110 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 049bb42 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/4/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/4/artifact/out/new-findbugs-hadoop-hdfs-project_hadoop-hdfs.html |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/4/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/4/testReport/ |
| Max. process+thread count | 4668 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

leosunli commented:
@goiri  I have updated this patch as your comments. The exists unit test is coverage in this area.Could you have time to help review this patch? Thank you.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 30 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1047 | trunk passed |
| +1 | compile | 62 | trunk passed |
| +1 | checkstyle | 41 | trunk passed |
| +1 | mvnsite | 70 | trunk passed |
| +1 | shadedclient | 758 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 53 | trunk passed |
| 0 | spotbugs | 157 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 156 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 60 | the patch passed |
| +1 | compile | 55 | the patch passed |
| +1 | javac | 55 | the patch passed |
| -0 | checkstyle | 38 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 2 new + 28 unchanged - 0 fixed = 30 total (was 28) |
| +1 | mvnsite | 60 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 712 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 42 | the patch passed |
| +1 | findbugs | 165 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 5728 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 47 | The patch does not generate ASF License warnings. |
| | | 9190 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.TestDatanodeDeath |
|   | hadoop.hdfs.TestReplication |
|   | hadoop.hdfs.TestReconstructStripedFileWithRandomECPolicy |
|   | hadoop.hdfs.server.diskbalancer.TestDiskBalancer |
|   | hadoop.hdfs.TestDFSStripedInputStreamWithRandomECPolicy |
|   | hadoop.hdfs.TestFileChecksumCompositeCrc |
|   | hadoop.hdfs.TestDFSStripedOutputStream |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux ed36d96eea0e 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a79bdf7 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/5/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/5/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/5/testReport/ |
| Max. process+thread count | 5279 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/5/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
+     * a thread for parallel MD5 computing to increase performance when loading
+     */
+    private static class DigestThread extends Thread {
+      volatile private IOException ioe = null;
```
Javadocs of the style:
/** Exception thrown when computing the diggest if it cannot be calculated. */
private volatile IOException ioe = null;
/** Calculated digest if there are no error. */
private volatile MD5Hash digest = null;

Note that I think the style says "private volatile"

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
+    private static class DigestThread extends Thread {
+      volatile private IOException ioe = null;
+      volatile private MD5Hash digest = null;
+      private File file;
```
final and add javadoc.

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
+      }
+
+      public MD5Hash getDigest() {
+        return digest;
```
What about checking if the exception is there and throw it if so?

leosunli replied:
if digest have exception ，DigestThread have catch it.
try {
          digest = MD5FileUtils.computeMd5ForFile(file);
        } catch (IOException e) {
          ioe = e;
        } catch (Throwable t) {
          ioe = new IOException(t);
        }

goiri replied:
Correct, but we could move all the exception handling there. 
My proposal is that the getter here checks if ioe == null and throw it.
Then it goes to my other comment where you don't need to check for the exception anymore and just do a getter of the digest.

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
+      private File file;
+
+      public DigestThread(File inFile) {
+        file = inFile;
```
It would be nice to have a name for the thread:
- setName
- toString (and let the daemon library do it if so)

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
     void load(File file) throws IOException {
       long start = Time.monotonicNow();
-      imgDigest = MD5FileUtils.computeMd5ForFile(file);
+      DigestThread dt = new DigestThread(file);
```
Should this be a daemon?

goiri replied:
Actually how expensive are these two steps? (digest calculation and load)

leosunli replied:
Test the patch against a fsimage of a 70PB  cluster (200million files and 300million blocks), the image loading time be reduced from 1210 seconds to 1105 seconds.

goiri replied:
That's fair, can you extend the description of the JIRA saying it can reduce up to 10% of the time or similar?

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
+          if (dt.getException() != null) {
+            throw dt.getException();
+          }
+          imgDigest = dt.getDigest();
```
If you do the approach of getDigest to throw the exception, you cover this.

leosunli replied:
this exception is throwed when call imgDigest = MD5FileUtils.computeMd5ForFile(file);
according to digest calculation and load code，this exception is also directly throwed. So I do it.

goiri replied:
My point is following the other comment.
If getDigest checks if there is an exception and throws it here, you don't need to check for the exception here.
You can also remove dt.getException().

leosunli commented:
@goiri I have updated this patch. Could you continue to help review this patch? Thank you.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 35 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1136 | trunk passed |
| +1 | compile | 63 | trunk passed |
| +1 | checkstyle | 44 | trunk passed |
| +1 | mvnsite | 67 | trunk passed |
| +1 | shadedclient | 763 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 48 | trunk passed |
| 0 | spotbugs | 177 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 173 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 65 | the patch passed |
| +1 | compile | 57 | the patch passed |
| +1 | javac | 57 | the patch passed |
| -0 | checkstyle | 35 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 2 new + 28 unchanged - 0 fixed = 30 total (was 28) |
| +1 | mvnsite | 63 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 710 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 48 | the patch passed |
| +1 | findbugs | 168 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 4888 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 39 | The patch does not generate ASF License warnings. |
| | | 8455 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.TestDatanodeDeath |
|   | hadoop.hdfs.server.datanode.TestDataNodeHotSwapVolumes |
|   | hadoop.hdfs.qjournal.server.TestJournalNodeSync |
|   | hadoop.hdfs.TestReconstructStripedFileWithRandomECPolicy |
|   | hadoop.hdfs.server.diskbalancer.TestDiskBalancer |
|   | hadoop.hdfs.server.namenode.ha.TestDFSUpgradeWithHA |
|   | hadoop.hdfs.TestDFSStorageStateRecovery |
|   | hadoop.hdfs.TestDFSStripedOutputStream |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/7/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 6fe5028fc878 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / be80334 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/7/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/7/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/7/testReport/ |
| Max. process+thread count | 4925 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/7/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 128 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1373 | trunk passed |
| +1 | compile | 79 | trunk passed |
| +1 | checkstyle | 48 | trunk passed |
| +1 | mvnsite | 84 | trunk passed |
| +1 | shadedclient | 986 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 65 | trunk passed |
| 0 | spotbugs | 192 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 189 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 79 | the patch passed |
| +1 | compile | 73 | the patch passed |
| +1 | javac | 73 | the patch passed |
| -0 | checkstyle | 44 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 2 new + 28 unchanged - 0 fixed = 30 total (was 28) |
| +1 | mvnsite | 84 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 906 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 59 | the patch passed |
| +1 | findbugs | 208 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 6774 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 35 | The patch does not generate ASF License warnings. |
| | | 11304 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |
|   | hadoop.hdfs.server.namenode.TestDecommissioningStatus |
|   | hadoop.hdfs.server.diskbalancer.TestDiskBalancer |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.server.datanode.TestDataNodeVolumeFailure |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.5 Server=18.09.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux e8a49be6f122 4.15.0-48-generic #51-Ubuntu SMP Wed Apr 3 08:28:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / be80334 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/6/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/6/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/6/testReport/ |
| Max. process+thread count | 2728 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/6/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,72 @@ public LoaderContext getLoaderContext() {
+
+      @Override
+      public String toString() {
+        return "DigestThread{" + "ThreadName=" + getName() + ", digest="
```
No need for the first +

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,72 @@ public LoaderContext getLoaderContext() {
       long start = Time.monotonicNow();
-      imgDigest = MD5FileUtils.computeMd5ForFile(file);
+      DigestThread dt = new DigestThread(file);
+      dt.setName(file.getName() + " computed MD5 Thread");
```
You can do this setter and the daemon one in the constructor of the class.
The text should be something like "file.getName() + " MD5 compute"
(No need for the word Thread specifically.)

leosunli commented:
@goiri  Thanks for your comments. I update this patch. Could you continue to review it? Thank you.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 69 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1112 | trunk passed |
| +1 | compile | 59 | trunk passed |
| +1 | checkstyle | 42 | trunk passed |
| +1 | mvnsite | 67 | trunk passed |
| +1 | shadedclient | 795 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 50 | trunk passed |
| 0 | spotbugs | 166 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 164 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 61 | the patch passed |
| +1 | compile | 56 | the patch passed |
| +1 | javac | 56 | the patch passed |
| -0 | checkstyle | 37 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 2 new + 28 unchanged - 0 fixed = 30 total (was 28) |
| +1 | mvnsite | 62 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 774 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 49 | the patch passed |
| +1 | findbugs | 168 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 6234 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 9904 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.tracing.TestTracing |
|   | hadoop.hdfs.TestMultipleNNPortQOP |
|   | hadoop.hdfs.server.namenode.ha.TestDFSUpgradeWithHA |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.5 Server=18.09.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/8/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux ec764a9e9631 4.15.0-48-generic #51-Ubuntu SMP Wed Apr 3 08:28:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 4a21224 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/8/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/8/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/8/testReport/ |
| Max. process+thread count | 2731 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/8/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

goiri commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java:
```diff
@@ -172,13 +172,72 @@ public LoaderContext getLoaderContext() {
       return ctx;
     }
 
+    private static class DigestThread extends Thread {
```
Javadoc explaining the overall idea. Something like:
Thread to compute the MD5 of a file as this can be in parallel while loading the image without interfering much.

Actually, this might be a good util to have in MD5FileUtils.

leosunli replied:
Thank @goiri   for reviewing this patch. Its purpose is  computing file md5 in parallel with image loading. So I think it might not be a good util to have in MD5FileUtils. Please correct me if I am wrong. Thanks.

leosunli commented:
@goiri  I have added Javadoc explaining the overall idea and updated patch.Could you help review this patch? If have no question, please merge it trunk.Thank you.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 140 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1762 | trunk passed |
| +1 | compile | 80 | trunk passed |
| +1 | checkstyle | 60 | trunk passed |
| +1 | mvnsite | 93 | trunk passed |
| +1 | shadedclient | 1040 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 66 | trunk passed |
| 0 | spotbugs | 202 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 199 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 70 | the patch passed |
| +1 | compile | 61 | the patch passed |
| +1 | javac | 61 | the patch passed |
| -0 | checkstyle | 50 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 2 new + 28 unchanged - 0 fixed = 30 total (was 28) |
| +1 | mvnsite | 77 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 887 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 52 | the patch passed |
| +1 | findbugs | 182 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 6291 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 34 | The patch does not generate ASF License warnings. |
| | | 11192 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.server.balancer.TestBalancerWithMultipleNameNodes |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.5 Server=18.09.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/9/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 2bc04a6e7af5 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / d203045 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/9/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/9/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/9/testReport/ |
| Max. process+thread count | 2700 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/9/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

goiri commented:
Can we fix the checkstyle warnings?

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 129 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1278 | trunk passed |
| +1 | compile | 65 | trunk passed |
| +1 | checkstyle | 46 | trunk passed |
| +1 | mvnsite | 71 | trunk passed |
| +1 | shadedclient | 844 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 52 | trunk passed |
| 0 | spotbugs | 170 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 168 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 64 | the patch passed |
| +1 | compile | 59 | the patch passed |
| +1 | javac | 59 | the patch passed |
| +1 | checkstyle | 40 | the patch passed |
| +1 | mvnsite | 65 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 811 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 50 | the patch passed |
| +1 | findbugs | 180 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 6893 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 48 | The patch does not generate ASF License warnings. |
| | | 10948 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.tools.TestDFSZKFailoverController |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.server.namenode.ha.TestDFSUpgradeWithHA |
|   | hadoop.hdfs.server.namenode.ha.TestHAAppend |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.5 Server=18.09.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/10/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1010 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 519877ca3588 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 2f4b37b |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/10/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/10/testReport/ |
| Max. process+thread count | 2579 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1010/10/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

leosunli commented:
Thank @goiri for your suggestion.  I have fixed the checkstyle warnings. Could you have time to review this patch ? Thank you.

goiri commented:
LGTM
+1

leosunli commented:
Hi @goiri, If you think this patch no problem, could you help merge it to trunk ? Thank you.

goiri commented:
Merged and updated the JIRA.
I'll update the patch there for completeness.
